scripts/image_builder.py

```
@@ -0,0 +1,803 @@
+            path = Path(out_name)
+        else:
+            path = PLO_SCRIPT_DIR / out_name
+    else:
+        path = PLO_SCRIPT_DIR / os.path.basename(script_name).removesuffix(".yaml")
+
+    with open(path, "w", encoding="ascii") as f:
+        progs = plo_script.write(f, enc)
+
+    # TODO: what about plo_script.size?
+    progs = [ProgInfo(path, plo_script.offs, os.path.getsize(path))] + progs
+    return progs
+
+
+
+def set_offset(file: IO[bytes], target_offset: int, padding_byte: int):
+    """Sets the file position to `target_offset` - write `padding_byte` to extend the file if necessary"""
+    assert 0 <= padding_byte <= 255
+    CHUNK_SIZE = 512
+
+    # always move to the end of the file (previous write might have been in the middle)
+    file.seek(0, os.SEEK_END)
+    curr_offset = file.tell()
+
+    if (diff := target_offset - curr_offset) > 0:
+
+        full, part = (diff // CHUNK_SIZE, diff % CHUNK_SIZE)
+        pad_chunk = bytes([padding_byte]) * CHUNK_SIZE
+        file.write(pad_chunk[:part])
+        for _ in range(full):
+            file.write(pad_chunk)
+    elif target_offset != curr_offset:  # overwriting parts of the existing image - just set the position directly
+        file.seek(target_offset)
+
+    assert file.tell() == target_offset, f"{file.tell()} != {target_offset}"
+
+
+def add_to_image(fout: BinaryIO, offset: int, fpath: Union[str, Path], padding_byte: int):
+    """Add file data (`fpath`) to the target image file (`fout`) at a given `offset`"""
+    set_offset(fout, offset, padding_byte)
+
+    CHUNK_SIZE = 512
+    written = 0
+    with open(fpath, "rb") as fIn:
+        while True:
+            data = fIn.read(CHUNK_SIZE)
+            written += fout.write(data)
+            if len(data) != CHUNK_SIZE:
+                break
+
+    return written
+
+
+def truncate(path: Union[str, Path], size: int):
+    try:
+        os.truncate(path, size)
+    except FileNotFoundError:
+        pass
```

Perhaps it would be better to crash the script instead of silently accepting that?

---

It's ok for the file to not exist (next `open(file, "w")` will create it (or fail) 

The `truncate(0)` is "lightweight" version of `rm -rf` (removes the contents without removing direntry).

This is actually a remnant of old script, I'll change it to `rm` (I don't want the file to exist if the script would fail with an error).